### ISSUE TITLE
NameError when ViewComponent source code preview is turned on

### ISSUE THREAD

**Issue by afomera** [allmarkedup/lookbook#11]
Hi there!

I pulled down the latest version of Lookbook and set it up like the following in my `development.rb` environment.

We're using `view_component` 2.35

```rb
  # ViewComponent preview configuration
  config.view_component.preview_paths << Rails.root.join("spec", "components", "previews")
  config.view_component.show_previews_source = true
  config.view_component.preview_route = "/view_component/previews"
  config.view_component.default_preview_layout = "view_component_preview"
```

Upon installing Lookbook per the Readme (add the gem from GitHub, add the mount line into routes and trying to view a basic view component preview (inside of Lookbook), I ran into this error:

<img width="1042" alt="CleanShot 2021-07-29 at 20 12 55@2x" src="https://user-images.githubusercontent.com/1741179/127589939-bd7d2761-c88c-4aba-b05a-2b49e75d0dc9.png">

We would love the ability to retain the ViewComponent source code previews in the native ViewComponent previews while Lookbook rapidly improves. 

Let me know if you need me to dive into this!

**Comment by allmarkedup** [comment]
Hey @afomera, thanks for posting this up.

I'm just looking into it, and it seems that I get the same error from the 'regular' view_component preview when using a custom `default_preview_layout` with `preview_source` helper in it.

Do you think you could do a quick test by removing Lookbook from the project altogether, keeping your view_component config the same and then just checking if it throws the same error when viewing the standard VC preview example URL? You may have already tried this -  apologies if so, just want to check if you are seeing the same thing as me.

Many thanks!

**Comment by allmarkedup** [comment]
@afomera Sorry, ignore my last message - I was just being an idiot! I've managed to recreate the issue now so will get it fixed as soon as I can.

**Comment by allmarkedup** [comment]
Hey @afomera, I've just released an update (v0.1.2) that I hope should fix the issue you were seeing. If you add lookbook to your Gemfile like this:

```ruby
gem "lookbook", '>= 0.1', git: "https://github.com/allmarkedup/lookbook", branch: "main"
```

And then run `bundle update` you will hopefully get the latest version.

Let me know if it fixes things for you - although I have to say I'm not sure why you would need the native ViewComponent source code display if you are using Lookbook? Would be interested to know why you want to leave that on - is it because you are also using the default ViewComponent preview rendering for something?

**Comment by afomera** [comment]
I did try this out this morning, but seem to have issues getting our app to boot at this point. I was on a pair with a coworker to show some of the lookbook stuff and had to revert back for the moment. I'll try and get more details soon 😅

We mostly wanted the view component source code stuff since we may need to link/show one component at a time.

**Comment by afomera** [comment]
Apologizes for taking so long,

<img width="1279" alt="CleanShot 2021-07-30 at 12 47 12@2x" src="https://user-images.githubusercontent.com/1741179/127698369-baaba95e-d2ac-4429-9093-77d5d50a2980.png">


Changing no configuration settings, and just updating our app with the config setting above, we receive this stack level too deep error.

**Comment by afomera** [comment]
Didn't change anything but was working elsewhere in our app, and now upon hitting Lookbook I get this:

<img width="1289" alt="CleanShot 2021-07-30 at 13 27 55@2x" src="https://user-images.githubusercontent.com/1741179/127702684-636877f1-7443-44b1-9ec0-f517d9fad0f3.png">

`code_object` in that case is nil.

**Comment by allmarkedup** [comment]
@afomera ok thank you - all good info. I'll try and dig into it some more as soon as I can, but might be a day or two now I'm afraid. Can you just confirm that you've run `bundle update` and have the latest version (0.1.2) showing in your Gemfile.lock?

Lookbook has not been battle tested with any large projects yet so there are likely to be more bugs like these - it's really great you are interested in using it but just to forewarn you :-)

Is the project you are testing this with open source? If so I'll try to pull it down and get it working and see if I can figure out what the issues are.

Very much appreciate you taking the time to dive into this, many thanks.

**Comment by afomera** [comment]
> @afomera ok thank you - all good info. I'll try and dig into it some more as soon as I can, but might be a day or two now I'm afraid. Can you just confirm that you've run bundle update and have the latest version (0.1.2) showing in your Gemfile.lock?

I do! Our gemfile looks like this:
`gem "lookbook", '>= 0.1.2', git: "https://github.com/allmarkedup/lookbook", branch: "main"`

```
GIT
  remote: https://github.com/allmarkedup/lookbook
  revision: 49186618e86bd4b65e7c729e5314bf8727611bc3
  branch: main
  specs:
    lookbook (0.1.2)
      actioncable
      listen
      rails
      redcarpet
      rouge
      view_component
      yard
```


> Lookbook has not been battle tested with any large projects yet so there are likely to be more bugs like these - it's really great you are interested in using it but just to forewarn you :-)

Yeah, totally understand that. Our team recognized that, but also accepted the risks with that. We're going to look at trying to contribute back as we can too (sorry it feels like I'm just filing issue after issue!) We've dug in a bit, but wanted to file issues in-case it was user-error or we were doing something wrong, I've been trying to get more comfortable with the gem code to try and fix / enhance some of these things but it takes me a bit to warm up :D

**Comment by allmarkedup** [comment]
@afomera sorry for the slow reply. I _think_ I've got a handle on the NoMethodError issue and will try and get a fix out for that as soon as I have a bit more time on my hands.

The SystemStackError error you are seeing is actually thrown by [Yard](https://github.com/lsegal/yard/) which is what Lookbook uses under the hood to parse out comments and tags in the preview files. So debugging that one might be a little tricker. This issue: https://github.com/lsegal/yard/issues/1116 might be of some use but I need to dig into it a bit more.

I'm going to be giving the parsing engine some attention soon as it's pretty rough and ready at the moment but whether that can fix the stack level error I'm not sure.

**Comment by afomera** [comment]
hey @allmarkedup!

Okay so I dug a little further into the SystemStackError and here's what I've come up with.

Small reminder of our setup for VC:

```
  config.view_component.preview_paths << Rails.root.join("spec", "components", "previews")
  config.view_component.show_previews_source = false
  config.view_component.preview_route = "/view_component/previews"
  config.view_component.default_preview_layout = "view_component_preview"
```

So, Lookbook uses the config.lookbook.listen_paths to loop over and parse for details/tags. (set https://github.com/allmarkedup/lookbook/blob/develop/lib/lookbook/engine.rb#L88) These get set as an array in the ```@paths``` instance variable in the Lookbook::Parser class

When I inspected `config.lookbook.listen_paths` it had both my custom config for `  config.view_component.preview_paths` and it also had `app/components` as well as set by https://github.com/allmarkedup/lookbook/blob/develop/lib/lookbook/engine.rb#L27-L29


I am not sure why it needs to parse the app/components directory for YARD tags when we should be adding them to the Preview classes honestly, so perhaps the fix is to setup a new config that lets you specify paths (and defaults to the view component preview_paths option? Perhaps there's a reason we want to be parsing the app/components directory?

For what it's worth, IDK what in our hundreds of components tripped this line, but when I binding.pry'd in and removed app/components from the `@paths` variable Lookbook magically loaded for me again. 

Here's a screenshot of my binding.pry experiment. 
<img width="1408" alt="CleanShot 2021-08-04 at 20 54 27@2x" src="https://user-images.githubusercontent.com/1741179/128283637-d2a63090-6790-4cb9-91a6-10df7f349de3.png">

I hope that helps! If a new config option is what you're thinking to solve this/work around we can do a PR as well adding that (or feel free to take a crack at this if this helped narrow things down?) just let me know! Thanks!

**Comment by allmarkedup** [comment]
Thanks @afomera that's very helpful :-)

To be honest, the only reason that `app/components` was included in the files that are parsed was to cover planned future functionality. For example, figuring out what component the preview is rendering, and then showing information about that component in the Lookbook UI (things like what params it accepts by parsing out `@param` tags, etc). It would also make sense in many cases to parse component-specific notes out of component files (rather than notes related to the preview specifically) and display them in the UI.

However... that functionality is probably some way off, and I'd rather get the preview/core implementation rock solid first. Parsing the component files (with all their possible complexities) is a much bigger surface area for problems to arise in so I think for now it would make sense to remove `app/components` from the parsing until there is a more concrete plan for that functionality, and there is a solid base to build off.

I think for now, I'll just remove the components directory from the parsing list, but not allow the parser paths to be configured just yet. Apart from adding in the components there is no real reason why someone might want to do that at this point, so I'd rather sit on that for a bit.

I'll try and make the change today and get a bump release out with the fix. Really appreciate you digging into this for me!

**Comment by afomera** [comment]
Great, thanks so much! That does make sense.

**Comment by allmarkedup** [comment]
Hey @afomera -  I've just released [v0.1.4](https://github.com/allmarkedup/lookbook/releases/tag/v0.1.4) - it simplifies the parsing process a lot (amongst other things) so I'm hoping this fixes the issues you've been having.

Could you give it a spin whenever you have time and see how you get on?

**Comment by allmarkedup** [comment]
@afomera I'm going to close this issue now as I believe it's fixed. Lookbook has now been [published as a a gem]9https://rubygems.org/gems/lookbook) (latest version v0.2.1) - if you run into any issues after updating to the latest version please just let me know :-)
